gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitConfigMonitor.java

```diff
@@ -0,0 +1,419 @@
+
+        String uri = this.git.getRepository().getConfig().getString("remote", "origin", "url");
+
+        if (!uri.equals(this.repoUri)) {
+          throw new RuntimeException("Repo at " + this.repoDir + " has uri " + uri + " instead of " + this.repoUri);
+        }
+      } catch (RepositoryNotFoundException e) {
+        // if the repository was not found then clone a new one
+        this.git = Git.cloneRepository()
+            .setDirectory(repoDirFile)
+            .setURI(this.repoUri)
+            .call();
+      }
+
+      try {
+        this.lastProcessedGitHash = readCheckpoint();
+      } catch (FileNotFoundException e) {
+        // if no checkpoint is available then start with the first commit
+        Iterable<RevCommit> logs = git.log().call();
+        RevCommit lastLog = null;
+
+        for (RevCommit log : logs) {
+          lastLog = log;
+        }
+
+        if (lastLog != null) {
+          this.lastProcessedGitHash = lastLog.getName();
+        }
+      }
+
+      this.latestGitHash = this.lastProcessedGitHash;
+    }
+
+    /**
+     * Read the last processed commit githash from the checkpoint file
+     * @return
+     * @throws IOException
+     */
+    private String readCheckpoint() throws IOException {
+      File checkpointFile = new File(this.repoDir, CHECKPOINT_FILE);
+      return Files.toString(checkpointFile, Charsets.UTF_8);
+    }
+
+    /**
+     * Write the last processed commit githash to the checkpoint file
+     * @param gitHash
+     * @throws IOException
+     */
+    private void writeCheckpoint(String gitHash) throws IOException {
+      // write to a temporary name then rename to make the operation atomic when the file system allows a file to be
+      // replaced
+      File tmpCheckpointFile = new File(this.repoDir, CHECKPOINT_FILE_TMP);
+      File checkpointFile = new File(this.repoDir, CHECKPOINT_FILE);
+
+      Files.write(gitHash, tmpCheckpointFile, Charsets.UTF_8);
+
+      Files.move(tmpCheckpointFile, checkpointFile);
+    }
+
+    private void moveCheckpointAndHashesForward() throws IOException {
+      this.lastProcessedGitHash = this.latestGitHash;
+
+      writeCheckpoint(this.latestGitHash);
+    }
+
+    /**
+     *
+     * @throws GitAPIException
+     * @throws IOException
+     */
+    private List<DiffEntry> getChanges() throws GitAPIException, IOException {
+      // get tree for last processed commit
+      ObjectId oldHeadTree = git.getRepository().resolve(this.lastProcessedGitHash + "^{tree}");
+
+      // refresh to latest
+      this.git.pull().call();
+      ObjectId head = this.git.getRepository().resolve("HEAD");
+      ObjectId headTree = this.git.getRepository().resolve("HEAD^{tree}");
+
+      // remember the hash for the current HEAD. This will be checkpointed after the diff is processed.
+      latestGitHash = head.getName();
+
+      // diff old and new heads to find changes
+      ObjectReader reader = this.git.getRepository().newObjectReader();
```

Comment:
No changes are made locally, so there should be no conflicts other than issues due to forced pushes going back in time.

I changed the code to do a hard reset instead of pull to avoid any issues.